File: examples/python/documentation/main-resources.py

from cdktf import App
# DOCS_BLOCK_START:resources-define
from constructs import Construct
from cdktf import TerraformStack
import imports.aws as aws

class HelloTerra(TerraformStack):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        aws.provider.AwsProvider(self, "aws",
            region = "us-east-1" 
        )

        region = aws.data_aws_region.DataAwsRegion(self, "region")

        aws.dynamodb_table.DynamodbTable(self, "Hello",
            name = "my-first-table-{}".format(region.name),
            hash_key = "temp",
            attribute = [{ "name": "id", "type": "S"}],
            billing_mode = "PAY_PER_REQUEST", 
        )
# DOCS_BLOCK_END:resources-define

# DOCS_BLOCK_START:resources-references
import imports.kubernetes as kubernetes 
# DOCS_BLOCK_END:resources-references

class TestStack(TerraformStack):
    def __init__(self, scope: Construct, name: str):
        super().__init__(scope, name)

        app = App()
        # DOCS_BLOCK_START:resources-references
        
        exampleNamespace = kubernetes.namespace.Namespace(self, "tf-cdk-example",
            metadata = kubernetes.namespace.NamespaceMetadata({
                "name": "tf-cdk-example"
            })
        )

        kubernetes.deployment.Deployment(self, "nginx-deployment",
            metadata = kubernetes.deployment.DeploymentMetadata({
                "name": "nginx",
                "namespace": exampleNamespace.metadata.name,
                "labels": {
                    app,
                }
            })
        )
        # DOCS_BLOCK_END:resources-references


class TestStack(TerraformStack):
    def __init__(self, scope: Construct, name: str):
        super().__init__(scope, name)
# DOCS_BLOCK_START:resources-escape-hatch
import imports.aws as aws
# DOCS_BLOCK_END:resources-escape-hatch

class TestStack(TerraformStack):
    def __init__(self, scope: Construct, name: str):
        super().__init__(scope, name)
        # DOCS_BLOCK_START:resources-escape-hatch
        
        tableName = "my-table"

        table = aws.dynamodb_table.DynamodbTable(self, "Hello",
            name = tableName,
            hash_key = "id",
            attribute = [{"name": "id", "type": "S"}]
        )

        table.add_override("provisioner", [
            {
                "local-exec": {
                    "command": "aws dynamodb create-backup --table-name {} --backup-name {}-backup".format(tableName)
                }
            }
        ])
        # DOCS_BLOCK_END:resources-escape-hatch

#DOCS_BLOCK_START:resources-override-attribute
import imports.aws as aws
#DOCS_BLOCK_END:resources-override-attribute


class TestStack(TerraformStack):
    def __init__(self, scope: Construct, name: str):
        super().__init__(scope, name)
        #DOCS_BLOCK_START:resources-override-attribute

        topic = aws.sns_topic.SnsTopic(self, "Topic",
            display_name = "will-be-overwritten"
        )

        topic.add_override("display_name", "my-topic")
        #DOCS_BLOCK_END:resources-override-attribute

#DOCS_BLOCK_START:resources-escape-hatch-dynamic-block
from cdktf import TerraformVariable
import imports.aws as aws
#DOCS_BLOCK_END:resources-escape-hatch-dynamic-block

class TestStack(TerraformStack):
    def __init__(self, scope: Construct, name: str):
        super().__init__(scope, name)
        #DOCS_BLOCK_START:resources-escape-hatch-dynamic-block
        
        ports = TerraformVariable(self, "ports",
            type = "list",
            default = [22, 80, 443, 5432] 
        )

        sq = aws.security_group.SecurityGroup(self, "sec1grp",
            name = "security1",
            vpc_id = "vpcs",
            egress = [
                {
                    "fromPort": 0,
                    "toPort": 0,
                    "ciderBlocks": ["0.0.0.0/0"],
                    "protocal": "-1"
                }
            ] 
        )

        sq.add_override("dynamic.ingress", {
            "for_each": ports.list_value,
            "content": {
                "fromPort": "${ingress.value}",
                "toPort": "${ingress.value}",
                "ciderBlocks": ["0.0.0.0/0"],
                "protocal": "-1"
            }
        })
        #DOCS_BLOCK_END:resources-escape-hatch-dynamic-block

#DOCS_BLOCK_START:resources-escape-loop-thru-ports
import imports.aws as aws
#DOCS_BLOCK_END:resources-escape-loop-thru-ports

class TestStack(TerraformStack):
    def __init__(self, scope: Construct, name: str):
        super().__init__(scope, name)
        #DOCS_BLOCK_START:resources-escape-loop-thru-ports
        
        ports = [22, 80, 443, 5432]

        aws.security_group.SecurityGroup(self, "sec1grp",
            name = "security1",
            vpc_id = "vpcs",
            egress = [
                {
                    "fromPort": 0,
                    "toPort": 0,
                    "ciderBlocks": ["0.0.0.0/0"],
                    "protocal": "-1"
                }
            ],
            ingress = map(lambda port_list, port: port_list.append(
                {
                    "fromPort": port,
                    "toPort": port,
                    "cidrBlocks": ["0.0.0.0/0"],
                    "protocol": "-1",
                }
            ), [], ports)
        )
        #DOCS_BLOCK_END:resources-escape-loop-thru-ports

File: examples/java/documentation/src/main/java/com/mycompany/app/resources/MainResources.java
package com.mycompany.app.resources;

// DOCS_BLOCK_START:resources-escape-hatch
// DOCS_BLOCK_START:resources-escape-hatch-dynamic-block
import com.hashicorp.cdktf.TerraformVariable;
import com.hashicorp.cdktf.TerraformVariableConfig;
// DOCS_BLOCK_END:resources-escape-hatch-dynamic-block
import imports.aws.dynamodb_table.DynamodbTable;
import imports.aws.dynamodb_table.DynamodbTableAttribute;
import imports.aws.dynamodb_table.DynamodbTableConfig;
// DOCS_BLOCK_END:resources-escape-hatch

// DOCS_BLOCK_START:resources-escape-hatch-dynamic-block
import imports.aws.security_group.*;
// DOCS_BLOCK_END:resources-escape-hatch-dynamic-block
import imports.kubernetes.deployment.*;
import software.constructs.Construct;
import com.hashicorp.cdktf.App;
import com.hashicorp.cdktf.TerraformStack;

// DOCS_BLOCK_START:resources-references
import imports.kubernetes.namespace.Namespace;
import imports.kubernetes.namespace.NamespaceConfig;
import imports.kubernetes.namespace.NamespaceMetadata;
// DOCS_BLOCK_END:resources-references
import imports.aws.provider.AwsProvider;
import imports.aws.provider.AwsProviderConfig;
import imports.kubernetes.provider.KubernetesProvider;
import imports.kubernetes.provider.KubernetesProviderConfig;
import imports.aws.sns_topic.SnsTopic;
import imports.aws.sns_topic.SnsTopicConfig;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class MainResources extends TerraformStack {

        public MainResources(Construct scope, String id) {
                super(scope, id);

                AwsProvider provider = new AwsProvider(this, "provider", AwsProviderConfig.builder()
                        .region("us-east-1")
                        .build()
                );
                new KubernetesProvider(this, "kind", KubernetesProviderConfig.builder()
                        .build()
                );
                App app = new App();
                // DOCS_BLOCK_START:resources-references
                Namespace exampleNamespace = new Namespace(this, "tf-cdk-example", NamespaceConfig.builder()
                                .metadata(NamespaceMetadata.builder()
                                                .name("tf-cdk-example")
                                                .build())
                                .build());

                new Deployment(this, "nginx-deployment", DeploymentConfig.builder()
                                .metadata(DeploymentMetadata.builder()
                                                .name("nginx")
                                                .namespace(exampleNamespace.getMetadata().getName()) // Reference the name property
                                                .labels(new HashMap<String, String>() {
                                                        {
                                                                put("app", app.toString());
                                                        }
                                                })
                                                .build())
                                .spec(DeploymentSpec.builder()
                                        .selector(DeploymentSpecSelector.builder()
                                                .matchLabels(new HashMap<String, String>(){{
                                                        put("app", app.toString());
                                                }})
                                                .build()
                                        )
                                        .replicas("1")
                                        .template(DeploymentSpecTemplate.builder()
                                                .metadata(DeploymentSpecTemplateMetadata.builder()
                                                        .labels(new HashMap<String, String>(){{
                                                                put("app", app.toString());
                                                        }})
                                                        .build()
                                                )
                                                .spec(DeploymentSpecTemplateSpec.builder()
                                                        .container(Arrays.asList(DeploymentSpecTemplateSpecContainer.builder()
                                                                        .image("nginx:1.7.9")
                                                                        .name("nginx")
                                                                        .build()
                                                                )
                                                        )
                                                        .build()
                                                )
                                                .build()
                                        )
                                        .build()
                                )
                                .build());
                // DOCS_BLOCK_END:resources-references

                // DOCS_BLOCK_START:resources-escape-hatch
                String tableName = "my-table";

                DynamodbTable table = new DynamodbTable(this, "Hello", DynamodbTableConfig.builder()
                                .name(tableName)
                                .hashKey("id")
                                .attribute(Arrays.asList(
                                                DynamodbTableAttribute.builder()
                                                                .name("id")
                                                                .type("S")
                                                                .build()))
                                .build());

                table.addOverride("provisioner", Arrays.asList(
                                new HashMap() {
                                        {
                                                put("local-exec", new HashMap<String, String>() {
                                                        {
                                                                put("command", "aws dynamodb create-backup --table-name "
                                                                                + tableName + " --backup-name "
                                                                                + tableName + "-backup");
                                                        }
                                                });
                                        }
                                }));
                // DOCS_BLOCK_END:resources-escape-hatch

                // DOCS_BLOCK_START:resources-escape-hatch-dynamic-block
                TerraformVariable ports = new TerraformVariable(this, "ports", TerraformVariableConfig.builder()
                                .type("list")
                                .defaultValue(Arrays.asList(22, 80, 443, 5432))
                                .build());

                SecurityGroup sq = new SecurityGroup(this, "sec1grp", SecurityGroupConfig.builder()
                                .name("security1")
                                .vpcId("vpcs")
                                .egress(Arrays.asList(
                                                SecurityGroupEgress.builder()
                                                                .fromPort(0)
                                                                .toPort(0)
                                                                .cidrBlocks(Arrays.asList("0.0.0.0/0"))
                                                                .protocol("-1")
                                                                .build()))
                                .build());

                sq.addOverride("dynamic.ingress", new HashMap<String, Object>() {
                        {
                                put("for_each", ports.getListValue());
                                put("content", new HashMap<String, Object>() {
                                        {
                                                put("from_port", "${ingress.value}");
                                                put("to_port", "${ingress.value}");
                                                put("cidr_blocks", Arrays.asList("0.0.0.0/0"));
                                                put("protocol", "-1");
                                        }
                                });
                        }
                });
                // DOCS_BLOCK_END:resources-escape-hatch-dynamic-block

                // DOCS_BLOCK_START:resources-escape-loop-thru-ports
                List<Number> myPorts = Arrays.asList(22, 80, 443, 5432);
                List<SecurityGroupIngress> ingress = new ArrayList();
                myPorts.forEach(port -> ingress.add(
                                SecurityGroupIngress.builder()
                                                .toPort(port)
                                                .fromPort(port)
                                                .cidrBlocks(Arrays.asList("0.0.0.0/0"))
                                                .protocol("-1")
                                                .build()));

                new SecurityGroup(this, "sec2grp", SecurityGroupConfig.builder()
                                .name("security1")
                                .vpcId("vpcs")
                                .egress(Arrays.asList(
                                                SecurityGroupEgress.builder()
                                                                .fromPort(0)
                                                                .toPort(0)
                                                                .cidrBlocks(Arrays.asList("0.0.0.0/0"))
                                                                .protocol("-1")
                                                                .build()))
                                .ingress(ingress)
                                .build());
                // DOCS_BLOCK_END:resources-escape-loop-thru-ports

                // DOCS_BLOCK_START:resources-override-attribute
                new SnsTopic(this, "Topic", SnsTopicConfig.builder()
                                .displayName("will-be-overwritten")
                                .build()).addOverride("display_name", "my-topic");
                // DOCS_BLOCK_END:resources-override-attribute

        }
}
